File: springwebfluxtutorial/src/main/java/com/qiusuo/webfluxtutorial/reactive/OptimizableOperator.java
package com.qiusuo.webfluxtutorial.reactive;

import reactor.core.CorePublisher;
import reactor.core.CoreSubscriber;
import reactor.util.annotation.Nullable;

public interface OptimizableOperator<IN, OUT> extends CorePublisher<IN> {

    //subscribeOrReturn is the real function that subscribes the subscriber.
    CoreSubscriber<? super OUT> subscribeOrReturn(CoreSubscriber<? super IN> actual) throws Throwable;

    CorePublisher<? extends OUT> source();

    @Nullable
    OptimizableOperator<?, ? extends OUT> nextOptimizableSource();
}

File: springwebfluxtutorial/src/main/kotlin/com/qiusuo/webfluxtutorial/reactive/OptimizableOperator.kt
package com.qiusuo.webfluxtutorial.reactive

import reactor.core.CorePublisher
import reactor.core.CoreSubscriber
import reactor.util.annotation.Nullable

interface OptimizableOperator<IN, OUT> : CorePublisher<IN> {
    //subscribeOrReturn is the real function that subscribes the subscriber.
    @Throws(Throwable::class)
    fun subscribeOrReturn(actual: CoreSubscriber<IN>): CoreSubscriber<IN>
    fun source(): CorePublisher<OUT>

    @Nullable
    fun nextOptimizableSource(): OptimizableOperator<IN, OUT>
}
